### ISSUE TITLE
SR/F (and SR/S) Logs have a new format with no official decode examples

### ISSUE THREAD

**Issue by briantrice** [zero-motorcycle-community/zero-log-parser#16]
Zero has a new format for SR/F logs which only roughly matches the existing formats, so the parser will almost need to be written from scratch.

To be clear on one point: we have no copies of officially decoded logs to emulate. This is key to understanding the data embedded (though the basic record format and messages are recoverable).

It is encoded in a more clever way which is taking time to develop against. Assistance would be most appreciated.

**Comment by hoegge** [comment]
Seems like the parser can partially understand the log from SR/S but not the time stand for example. I have looked at the code, and for an outsider it seems like to big a challenge to try to decode it. Can I help, if I can supply both binary and decoded examples for you?

**Comment by atomicdog** [comment]
A decoded file with the binary would be very helpful. Without a decoded file there is only a small amount of data that can be decoded from the binary and it's a very tedious process.

**Comment by briantrice** [comment]
Thanks for the feedback. Yes, please provide decodes if at all possible. My Google ID is briantrice at gmail dot com and you may email me or use Google Drive to share confidential files.

My policy is to treat all submitted information as confidential, and if I add derived information to a test suite, I will ensure that you've given permission to use it, and I'll do my best to scrub personal or identifiable information (usually just VIN and perhaps configuration or ride details depending on context).

**Comment by briantrice** [comment]
Pull requests, patches, source file suggestions, etc. are all welcome. I'll do my best to accept whatever seems valid and safe.

I will say that it's clear that the firmware on the bike produces the log events in a binary format, directly translated from internal binary state machines, as updated from binary CAN messages, and does not include a fully-featured log event printer, so Zero is being very jealous with interpretation of the motorcycle's operation.

Now, I am not a lawyer (IANAL), but I would argue that if an owner of this generation of Zero motorcycle had a dispute with Zero over the operation of the bike, Zero has decided to privilege their own corporate access over the interests of their owners. Owners have currently no way to dispute or even see Zero's interpretation of the bike's logs.

I believe that laws concerning Right to Repair have something to say about this, and I invite owners to look into this on their own for whatever legal jurisdiction they live in.

**Comment by atomicdog** [comment]
I got distracted by other projects, but I had started a [fork](https://github.com/atomicdog/zero-log-parser) where I'm rewritten the parser using [Kaitai struct](https://kaitai.io/#what-is-it).  

It looks like the serial console commands also display data that's at least close to the data' order in the logs.
For example, the charger info displayed looks to be in the same order as the log data.

**Comment by briantrice** [comment]
That's a great idea. I would like you to contact me privately about your development process and lessons learned, if you please. My GitHub username at gmail is best.

**Comment by briantrice** [comment]
I've made a PR representing your changes. Input or requests welcome. I'll see whether I can test it soon.

**Comment by kscholty** [comment]
Hi, what's the status of this? Seems quite dead by now...
Is there any hope that we will get a parser for the new models? My DSR/X logs produce a lot of gibberish....

**Comment by atomicdog** [comment]
It's a 'tough nut to crack'. I work on it from time to time but the lack of progress gets frustrating, so I haven't tried recently.

**Comment by kscholty** [comment]
Yes, I can see that. If there is anything I can do to help, let me know.

**Comment by cecil-t** [comment]
I've been going through SR/F era files and have a few things that might be useful. All numbers here are from the current main, commit ff7f770.

First, a correction I think matters more than the decode work. LOG_STRUCTURE.md describes the ring buffer format as 2024 firmware and compressed telemetry as 2025+ firmware. In my data these track the bike platform, not the firmware year. Position 4 of the VIN is the platform character in Zero's own VIN tables (from the owner's manuals): S is the SDS platform (S, DS, SR, DSR), X is the FX line, and Z and D are the two halves of what Zero calls FST, street and adventure. Every file in the newer layouts is Z or D, every file in the legacy layouts is S or X, across 7,492 files with zero overlap. Most of my compressed-telemetry files are model year 2020, which is when the SR/F launched, and the legacy 262,144 byte format is still showing up on S and X bikes as late as 2024. So the two ran side by side for years (and maybe still) rather than one replacing the other.

I can say these layouts existed in 2020 because I have the files. I don't have any 2025 or 2026 files in those classes, so I can't say anything useful about the recent end of it.

Second, where the parser actually stands on FST platform bikes. On the MBB side, 418 files and 827,005 entries decoded, 95.93% to known types. Legacy platform MBB is 3,635 files at 99.99%, so the newer platform is close but not there. On the BMS side it's 77% known, which is the bigger gap of the two.

The MBB gap looks closeable. I have 33,665 unknown entries on this platform, and six type codes account for 97.75% of them: 0x48 (9,478 entries), 0x52 (7,216), 0x53 (5,837), 0x4B (5,406), 0x4F (2,804) and 0x49 (2,165). Everything else is 758 entries spread across six more codes. So this isn't a long tail, it's six specific messages with thousands of examples each. Decoding those would put FST MBB decodes at roughly 99.9%, matching legacy.  Only 0x52 appears in LOG_STRUCTURE.md, as Unknown Type 82. 0x4F has logQ, CAN1txQ and CAN2txQ as ASCII tags so it looks like queue diagnostics, and 0x48 has Charger in every example I've looked at. The other four are dense binary.

Two smaller things. The 131,200 and 131,328 byte files are different layouts, not one format with a different trailer. All 290 of my 131,328 files are built from 128-byte pages each starting with 00 f0 ff 00 at every page boundary, no exceptions, and that marker appears nowhere in the 453 files at 131,200. Also, the 34 files I have at exactly 262,143 bytes are just ring buffer files truncated by one byte, and they parse fine at 100% known types.

We can discuss sending my data set if you need it, let me know.

**Comment by atomicdog** [comment]
`0x48`  Charger module details
`0x49`  Charging state
`0x4B - 0x4D` telemetry snapshots (small/medium/large)
`0x51 - 0x53`  telemetry snapshots (small/medium/large)
`0x4F` Shows the fault when a write to a full FreeRTOS Queue happens.

**Comment by atomicdog** [comment]
@cecil-t I agree the LOG_STRUCTURE.md should have the different log files separated by platform not year.
The 131,328 byte files should all be BMS files and the the smaller one MBB's.  The  `00 f0 ff 00` looks like data corruption where a firmware bug overwrites, on the page boundary, the data that should have been there.

> We can discuss sending my data set if you need it, let me know.

I'm interested in getting your  dataset, I only have a few log files (mostly my bike) to analysis.

**Comment by cecil-t** [comment]
@atomicdog appreciate you digging into this. I ran both of your comments against the full file set (7565 files) to check.

On the size split: close but not universal. 82% of the Gen3 MBB files land right at 131,200 bytes, exactly 128 bytes under BMS's fixed 131,328, so your read holds for most of them. Two files come in larger though, at 261,760 bytes, both sharing a distinct _ecuId[N].bin filename pattern, so "smaller = MBB" isn't a clean rule across the board.

On 00 f0 ff 00: it's fairly precisely structured, shows up at a consistent 128-byte stride inside the 131,328-byte BMS format specifically (well over 99% of every occurrence lands there). But checked against actual flash geometry (256-byte write pages, 4096-byte erase sectors) it doesn't line up with either, so I see it as a page marker internal to this one log format rather than a boundary artifact.

Could use another set of eyes on the data set. Feel free to email me (in my profile) to discuss further.
